Block data mover design (#9528)
vmware-tanzu/velero: design/block-data-mover/block-data-mover.md

**Exposer**: While the restorePV is in block mode, exposer needs to rebind the restorePV to a targetPVC in either file system mode or block mode.    
**Uploader**: The same block uploader holds special logics to make performant data write to block devices and holds special logics to read  data from the backup chain in Unified repository.  

## Detailed Design

### Selectable Data Mover Type
At present, the backup accepts a `DataMover` parameter and when its value is empty or `velero`, VBDM is used.  
After block data mover is introduced, VBDM will have two types of data movers, Velero file system data mover and Velero block data mover.  
A new type string `velero-block` is introduced for Velero block data mover, that is, when `DataMover` is set as `velero-block`, Velero block data mover is used.  
Another new value `velero-fs` is introduced for Velero file system data mover, that is, when `DataMover` is set as `velero-fs`, Velero file system data mover is used.  
For backwards compatibility consideration, `velero` is preserved a valid value, it refers to the default data mover, and the default data mover may change among releases.  

> When `DataMover` is empty or `velero`, the default data mover is used. At present, Velero file system data mover is the default data mover; we can change the default one to Velero block data mover in future releases.  

Author: Yes, it is supported. I've added the elaboration into the design.

**Velero Built-in Data Mover (VBDM)**: VBDM, which is introduced in [Volume Snapshot Data Movement design][2] and [Unified Repository design][1], is the built-in data mover shipped along with Velero, it includes Velero data mover controllers and VGDP.  
**Data Mover Pods**: Intermediate pods which hold VGDP and complete the data transfer. See [VGDP Micro Service for Volume Snapshot Data Movement][3] for details.   
**Change Block Tracking (CBT)**: CBT is the mechanism to track changed blocks, so that backups could back up the changed data only. CBT usually provides by the comupting/storage plateform.  
**TCO**: Total Cost of Ownership. This is a general criteria for products/solutions, but also means a lot for BR solutions. For example, this means what kind of backup storage (and its cost) it requires, the retention policy of backup copies, the ways to remove backup data redundancy, etc.  

## Background
At present, [CSI Snapshot Data Movement][2] which is implemented by VBDM, ships a file system uploader, so the backup/restore is done from file system only.  
However, once possible, block level backup/restore is better than file system level backup/restore:  

> - Block level backup could leverage CBT to process minimual size of data so significantly reduce the overhead to network, backup repository and backup storage. As a result, TCO is significantly reduced.

Reviewer: ```suggestion
- Block level backup could leverage CBT to process minimal size of data to significantly reduce the overhead to network, backup repository and backup storage. As a result, TCO is significantly reduced.
```

- Block level backup/restore is performant in throughput and resource consumption, because it doesn't need to handle the complexity of the file system, especially for the case that huge number of small files in the file system.
- Block level backup/restore is less OS dependent because the uploader doesn't need the OS to be aware of the file system in the volume.

> At present, [Kubernetes CBT API][4] is mature and close to Beta stage. Many plateform/storage has supported/is going to support it.  

Reviewer: ```suggestion
At present, [Kubernetes CBT API][4] is mature and close to Beta stage. Many platform/storage has supported/is going to support it.  
```

Author: Done

Therefore, it is very important for Velero to deliver the block level backup/restore and recommend users to use it over the file system data mover as long as:  
- The volume is backed by block storage so block level access is possible

> - The plateform supports CBT

Reviewer: ```suggestion
- The platform supports CBT
```

Here are the data structs for VolumePolicy:  
```go
type volPolicy struct {
	action     Action
	conditions []volumeCondition
}

type volumeCondition interface {
	match(v *structuredVolume) bool
	validate() error
}

type structuredVolume struct {
	capacity     resource.Quantity
	storageClass string
	nfs          *nFSVolumeSource
	csi          *csiVolumeSource
	volumeType   SupportedVolume
	pvcLabels    map[string]string
	pvcPhase     string
}

type Action struct {
	Type VolumeActionType `yaml:"type"`
	Parameters map[string]any `yaml:"parameters,omitempty"`
}

const (
	ConfigmapRefType string = "configmap"
	Skip VolumeActionType = "skip"
	FSBackup VolumeActionType = "fs-backup"
	Snapshot VolumeActionType = "snapshot"
)
```

`action.parameters` is used to provide extra information of the action. This is an ideal place to differentate Velero file system data mover and Velero block data mover.  
Therefore, Velero built-in data mover will support `dataMover` key in `parameters`, with the value either `velero-fs` or `velero-block`. While `velero-fs` and `velero-block` are with the same meaning with Per Backup Selection.  

> Then, here is how users use this solution:  

Reviewer: ```suggestion
As an example, here is how a user might use both `velero-block` and `velero-fs` in a single backup:  
```

- Users set `DataMover` parameter for the backup as `velero-block`
- Users add a record into Volume Policy, make `conditions` to filter the volumes they want to backup through Velero file system data mover, make `action.type` as `snapshot` and insert a record into `action.parameter` as `dataMover:velero-fs`

In this way, all volumes matched by `conditions` will be backed up with Velero file system data mover; while the others will fallback to the per backup method Velero block data mover.  

Vice versa, users could set the per backup method as file system data mover and select volumes for Velero block data mover.  

The selected data mover for each volume should be recorded to `volumeInfo.json`.  

### Controllers
Backup controller and Restore controller are kept as is, async operations are still used to interact with VBDM with block data mover.  

> DataUpload controller and DataDownload controller are almost kpet as is, with some minor changes to handle the data mover type and backup type appropriately and convey it to the exposers. With [VGDP Micro Service][3], the controllers are almost isolated from VGDP, so no major changes are required.  

Author: Done

### Exposer

#### CSI Snapshot Exposer
The existing CSI Snapshot Exposer is reused with some changes to decide the backupPVC volume mode by access mode. Specifically, for Velero block data mover, access mode is always `Block`, so the backupPVC volume mode is always `Block`.  
Once the backupPVC is created with correct volume mode, the existing code could create the backupPod and mount the backupPVC appropriately.  

#### Generic Restore Exposer
The existing Generic Restore Exposer is reused, but the workflow needs some changes.  
For block data mover, the restorePV is in Block mode all the time, whereas, the targetPVC may be in either file system mode or block mode.  
However, Kubernetes doesn't allow to bound a PV to a PVC with mismatch volume mode.  

Therefore, the workflow of ***Finish Volume Readiness*** as introduced in [Volume Snapshot Data Movement design][2] is changed as below:
- When restore completes and restorePV is created, set restorePV's `deletionPolicy` to `Retain`

> - Create another rebindPV and copy restorePV's `volumeHandle`

Author: Done

For backwards compatibility consideration, `velero` is preserved a valid value, it refers to the default data mover, and the default data mover may change among releases. At present, Velero file system data mover is the default data mover; we can change the default one to Velero block data mover in future releases.  

#### Volume Policy
It is a valid case that users have multiple volumes in a single backup, while they want to use Velero file system data mover for some of the volumes and use Velero block data mover for some others.  
To meet this requirement, a combined solution of Per Backup Selection and Volume Policy is used.

Here are the data structs for VolumePolicy:  
```go
type volPolicy struct {
	action     Action
	conditions []volumeCondition
}

type volumeCondition interface {
	match(v *structuredVolume) bool
	validate() error
}

type structuredVolume struct {
	capacity     resource.Quantity
	storageClass string
	nfs          *nFSVolumeSource
	csi          *csiVolumeSource
	volumeType   SupportedVolume
	pvcLabels    map[string]string

> 	pvcPhase     string

Reviewer 2: Is it possible to use volumeMode and AccessMode as a filter? Thank you. This would be useful for our downstream case.

type ObjectWriteOptions struct {
	FullPath     string 	// Full logical path of the object
	DataType     int    	// OBJECT_DATA_TYPE_*
	Description  string 	// A description of the object, could be empty
	Prefix       ID     	// A prefix of the name used to save the object
	AccessMode   int    	// OBJECT_DATA_ACCESS_*
	BackupMode   int    	// OBJECT_DATA_BACKUP_*
	AsyncWrites  int    	// Num of async writes for the object, 0 means no async write
	ParentObject ID     	// the parent object based on which incremental write will be done
}
```

To support non-Kopia uploader to save snapshots to Unified Repo, snapshot related methods will be added to `BackupRepo` interface:
```go
	// SaveSnapshot saves a repo snapshot
	SaveSnapshot(ctx context.Context, snapshot Snapshot) (ID, error)

	// GetSnapshot returns a repo snapshot from snapshot ID
	GetSnapshot(ctx context.Context, id ID) (Snapshot, error)

	// DeleteSnapshot deletes a repo snapshot
	DeleteSnapshot(ctx context.Context, id ID) error

	// ListSnapshot lists all snapshots in repo for the given source (if specified)
	ListSnapshot(ctx context.Context, source string) ([]Snapshot, error)
```

> To support non-Kopia uploader to save metadata, which is used to describe the the backed up objects, some metadata related methods will be added to `BackupRepo` interface:

Reviewer: ```suggestion
To support non-Kopia uploader to save metadata, which is used to describe the backed up objects, some metadata related methods will be added to `BackupRepo` interface:
```

- File system IO to disk general aligned to a specific size, e.g., 4KB for NTFS and ext4, as long as the chunk size is a multiply of this size, it effectively reduces the case that one IO kills two deduplication chunks
- For the usage cases that the disk is used as raw block device without a file system, the IO is still conducted by aligning to a specific boundary

The chunk size is intentionally chosen as 1MB, reasons:
- 1MB is a multiply of 4KB for file systems or common block sizes for raw block device usages
- 1MB is the start boundary of partitions for modern operating systems, for both MBR and GPT, so partition metadata could be isolated to a separate chunk
- The more chunks are there, the more indexes in the repository, 1MB is a modreate value regarding to the overhead of indexes for Kopia repository

#### Benefits
Since the existing block address table(BAT) of CAOS is reused and kept as is, it brings below benefits:
- All the entries are still managed by Kopia CAOS, so Velero doesn't need to keep an extra data

> - The objects written by Velero block uploader is still recgnozible by Kopia, for both full backup and incremental backup

Reviewer: ```suggestion
- The objects written by Velero block uploader is still recognizable by Kopia, for both full backup and incremental backup
```

- The existing data management in Kopia repository still works for objects generated by Velero block uploader, e.g., snapshot GC, repository maintenance, etc.

Most importantly, this solution is super perfromant:
- During incremental write, it doesn't copy any data from the parent object, instead, it only clones object block address entries
- During backup deletion, it doesn't need to move any data, it only deletes the BAT for the object  

#### Uploader behavior
The block uploader's skippable write must also be aligned to this 1MB boundary, because Incremental Aware Object Extension needs to clone the entries that have been skipped from the parent object.   

> File system uploader is still using variable-sized deduplication, it is fine to keep data from the two uploaders into the same Kopia repository, though normally they won't be mutrally deduplicated.  

Reviewer: ```suggestion
File system uploader is still using variable-sized deduplication, it is fine to keep data from the two uploaders into the same Kopia repository, though normally they won't be mutually deduplicated.  
```

Author: Done

When volume resize happens, block uploader needs to handle it appropriately in below ways:
- Loop with CBT to the boundary of RoundDownTo1M(newSize)
- Read data between RoundDownTo1M(newSize) and newSize to get the tail data
- Call `WriteAt` from offset RoundDownTo1M(newSize) with the tail data

That is to say, the tail must be rewrite since Incremental Aware Object Extension cannot copy BAT entries halfly for the incremental backup.  

### Cancellation
The existing Cancellation mechanism is reused, so there is no change outside of the block uploader.  
Inside the uploader, cancellation checkpoints are embedded to the uploader reader and writer, so that the execution could quit in a reasonable time once cancellation happens.  

### Parallelism
Parallelism among data movers will reuse the existing mechanism --- load concurrency.  
Inside the data mover, uploader reader and writer are always running in parallel. The number of reader and writer is always 1.  
Sequaltial read/write of the volume is always optimized, there is no prove that multiple readers/writers are beneficial.  

### Progress Report
Progress report outside of the data mover will reuse the existing mechanism.  

> Innside the data mover, progress update is embedded to the uploader writer.  

Reviewer: ```suggestion
Inside the data mover, progress update is embedded to the uploader writer.  
```

#### Volume Policy
It is a valid case that users have multiple volumes in a single backup, while they want to use Velero file system data mover for some of the volumes and use Velero block data mover for some others.  
To meet this requirement, a combined solution of Per Backup Selection and Volume Policy is used.

Here are the data structs for VolumePolicy:  
```go
type volPolicy struct {
	action     Action
	conditions []volumeCondition
}

type volumeCondition interface {
	match(v *structuredVolume) bool
	validate() error
}

type structuredVolume struct {
	capacity     resource.Quantity
	storageClass string
	nfs          *nFSVolumeSource
	csi          *csiVolumeSource
	volumeType   SupportedVolume
	pvcLabels    map[string]string
	pvcPhase     string
}

type Action struct {
	Type VolumeActionType `yaml:"type"`
	Parameters map[string]any `yaml:"parameters,omitempty"`
}

const (
	ConfigmapRefType string = "configmap"
	Skip VolumeActionType = "skip"
	FSBackup VolumeActionType = "fs-backup"
	Snapshot VolumeActionType = "snapshot"
)
```

> `action.parameters` is used to provide extra information of the action. This is an ideal place to differentate Velero file system data mover and Velero block data mover.  

Reviewer: ```suggestion
`action.parameters` is used to provide extra information of the action. This is an ideal place to differentiate Velero file system data mover and Velero block data mover.  
```

- `GetMetadataDelta` is called with `BaseSnapshotId` set as the preserved changeId
- When deleting a backup, a VolumeSnapshot-VolumeSnapshotContent pair is rebuilt with `deletionPolicy` as `delete` and `snapshotHandle` as the preserved one
- Then the rebuilt VolumeSnapshot is deleted so that the volume snapshot is deleted from the storage

There is no way to automatically detect which way a specific volume support, so an interface is exposed to users to set the volume snapshot retention method.  
The interface could be added to the `Action.Parameters` of Volume Policy. By default, Velero block data mover takes Way 1, so volume snapshot is never retained; if users specify `RetainSnapshot` parameter, Way 2 will be taken.  
```go
type Action struct {
	Type VolumeActionType `yaml:"type"`
	Parameters map[string]any `yaml:"parameters,omitempty"`
}
```
In this way, users could specify --- for storage class "xxx" or CSI driver "yyy", backup through CSI snapshot with Velero block data mover and retain the snapshot.   

#### Incremental Size

> By the end of the backup, incremental size is also returned by the uploader, as same as Velero file system uploader. The size indicates how much data are unquie so processed by the uploader, based on the provided CBT.  

Author: Done

### Fallback to Full Backup
There are some occasions that the incremental backup won't continue, so the data mover fallbacks to full backup:
- `GetMetadataAllocated` or `GetMetadataDelta` returns error
- ChangeId is missing
- Parent snapshot is missing

### Irregular Volume Size
As mentioned above, during incremental backup, block uploader IO should be restricted to be aligned to the deduplication chunk size (1MB); on the other hand, there is no hard limit for users' volume size to be aligned.  
To support volumes with irregular size, below measures are taken:
- Volume objects in the repository is always aligned to 1MB
- If the volume size is irregular, zero bytes will be padded to the tail of the volume object
- A real size is recorded in the repository snapshot

> - During restore, the real size of data is restored

Author: Done.

The padding must be always with zero bytes.  

### Volume Size Change
Incremental backup could continue when volume is resized.  
Block uploader supports to write disk with arbibrary size.  
The volume resize cases don't need to be handled case by case.  

Instead, when volume resize happens, block uploader needs to handle it appropriately in below ways:
- Loop with CBT
- Read data between RoundDownTo1M(newSize) and newSize to get the tail data
- If there is no tail data, which means the volume size is aligned to 1MB, then call `WriteAt(newSize, nil)`
- Otherwise, call `WriteAt(RoundDownTo1M(newSize), taildata)`, `taildata` is also padded to 1MB

That is to day:

> - If CBT covers the tail of the volume, loop with CBT is enough for both shink and expand case

Author: Done

Backup type will also be added to `volumeInfo.json` to support observability purposes.  

Backup TTL is still used for users to specify a backup's retention time. By default, both full and incremental backups are with 30 days retention, even though this is not so reasonable for the full backups. This could be enhanced when Velero supports sophisticated retention policy.  
As a workaround, users could create two schedules for the same scope of backup, one is for full backups, with less frequency and longer backup TTL; the other one is for incremental backups, with normal frequency and shorter backup TTL.  

#### File System Data Mover
At present, Velero file system data mover doesn't support selectable backup type, instead, incremental backups are always conducted once possible.  
From user experience this is not reasonable.  

Therefore, to solve this problem and to make it align with Velero block data mover, Velero file system data mover will support backup type as well.  

At present, the data path for Velero file system data mover has already supported it, we only need to expose this functionality to users.  

### Backup Describe

> Backup type should be added to backup description, this value could be retrieved from `volumeInfo.json`.  

Reviewer: ```suggestion
Backup type should be added to backup description, this value could be retrieved from `volumeInfo.json`.   This would allow user to know if last backup was full, or incremental, or fallback to full.
```

Author: Done

There are rich features delivered with VGDP, VBDM and [VGDP micro service][3], to reuse these features, block data mover should be built based on these modules.  

Velero VBDM supports linux and Windows nodes, however, Windows container doesn't support block mode volumes, so backing up/restoring from Windows nodes is not supported until Windows container removes this limitation. As a result, if there are both linux and Windows nodes in the cluster, block data mover can only run in linux nodes.  

Both the Kubernetes CBT service and Velero work in the boundary of the cluster, even though the backend storage may be shared by multiple clusters, Velero can only protection workloads in the same cluster where it is running.  

## Goals

Add a block data mover to VBDM and support block level backup/restore for [CSI Snapshot Data Movement][2], which includes:  
- Support block level full backup for both `FileSystem` and `Block` mode volumes
- Support block level incremental backup for both `FileSystem` and `Block` mode volumes
- Support block level restore from full/incremental backup for both `FileSystem` and `Block` mode volumes

> - Support block level backup/restore for both linux and Windows workloads

Reviewer 3: Goals say "Support block level backup/restore for both linux and Windows workloads" but non-goals say "Backing up/restoring from Windows nodes is not supported." Can we clarify this? I think the intent is that Windows workload volumes can be backed up from a linux data mover pod, but the data mover pod itself won't run on Windows nodes. The current wording reads contradictory.

Author: Sure, I have changed the goal as `Support block level backup/restore for both linux and Windows workloads from linux cluster nodes`

**Uploader**: A new block uploader is added. It interacts with CBT layer, holds special logics to make performant data read from block devices and holds special logics to write incremental data to Unified Repository.  
**Extended Kopia repo**: A new Incremental Aware Object Extension is added to Kopia's CAOS, so as to support incremental data write. Other parts of Kopia repository, including the existing CAOS and CABS, are not changed.    

### Restore

Below is architecture for block data mover restore which is developed based on the existing VBDM:    

![Restore architecture](restore-architecture.png)

The existing VBDM is reused, below are the major changes based on the existing VBDM:  
**Exposer**: While the restorePV is in block mode, exposer needs to rebind the restorePV to a targetPVC in either file system mode or block mode.    
**Uploader**: The same block uploader holds special logics to make performant data write to block devices and holds special logics to read  data from the backup chain in Unified repository.  

For more details of VBDM, see [Volume Snapshot Data Movement design][2].  

## Detailed Design

### Selectable Data Mover Type

> #### Per Backup Selection

Reviewer 3: What happens if a user sets --data-mover velero-block but CBT isn't available for a volume? The design covers fallback from incremental to full, but does it also fall back from velero-block to velero-fs? Or does it attempt a full block backup without CBT? A full block backup without CBT on a 1TB volume means reading all 1TB which could be worse than the fs mover. Should we define the fallback chain clearly -- e.g. incremental block -> full block -> fs mover, or is it just incremental block ->full block -> error?

Author: We cannot fallback block data mover to fs data mover, because they are two different methods and different from the very beginning.

For block data mover backup, fallback won't happen frequently, it is only as a exception handling, so we don't need to over consider it.
And even it happens, our mechanisms in backup repo and uploader could optimize the full read/write.
Finally, even for the full read/write from block level itself, it is very fast, because it doesn't need to handle the file system complexities.
So it is not necessary to fallback to fs data mover automatically when CBT is just with some problem in seldom cases.

I have updated the design to clarify the fallback behavior.

	io.Seeker

	// Checkpoint is periodically called to preserve the state of data written to the repo so far.
	// Checkpoint returns a unified identifier that represent the current state.
	// An empty ID could be returned on success if the backup repository doesn't support this.
	Checkpoint() (ID, error)

	// Result waits for the completion of the object write.
	// Result returns the object's unified identifier after the write completes.
	Result() (ID, error)
}
```

To clone data from parent object, the caller needs to specify the parent object. To support this, `ObjectWriteOptions` is extended with `ParentObject`.  
The existing `AccessMode` could be used to indicate the data access type, either file system or block:  

```go
// ObjectWriteOptions defines the options when creating an object for write
type ObjectWriteOptions struct {
	FullPath     string 	// Full logical path of the object
	DataType     int    	// OBJECT_DATA_TYPE_*
	Description  string 	// A description of the object, could be empty
	Prefix       ID     	// A prefix of the name used to save the object
	AccessMode   int    	// OBJECT_DATA_ACCESS_*

> 	BackupMode   int    	// OBJECT_DATA_BACKUP_*

Reviewer 3: BackupMode field is added to ObjectWriteOptions but I don't see it explained anywhere in the design. What values does it take and what does it control?

Author: `BackupMode` is an existing field that was embedded to support full/incremental backup switch. This is NOT added in this design, we may or may not use it.

`ObjectWriteOptions` is an existing field and is already used, it is not for wider and more general purposes.

As another fact, directly saving `BlockMetadata` list will be memory consuming.  

With all the above considerations, the `Bitmap` data structure is used to save the allocated/changed blocks, calling CBT Bitmap.  
CBT Bitmap chunk size could be set as 1MB or a multiply of it, but a larger chunk size would amplify the backup size, so 1MB size will be use.  

Finally, interactions among CSI Snapshot Metadata Service, CBT Layer and Uploader is like below:  

![CBT Layer](cbt.png) 

In this way, CBT layer and uploader are decoupled and CBT bitmap plays as a north bound parameter of the uploader.  

#### Block Uploader
Block uploader consists of the reader and writer which are running asynchronously.  
During backup, reader reads data from the block device and also refers to CBT Bitmap for allocated/changed blocks; writer writes data to the Unified Repo.
During restore, reader reads data from the Unified Repo; writer writes data to the block device.  

> Reader and writer connects by a ring buffer, that is, reader pushes the block data to the ring buffer and writer gets data from the ring buffer and write to the target.  

Reviewer 3: On restarts, the block uploader has a ring buffer, reader/writer state, and position in the CBT bitmap. On restart, does Checkpoint() on ObjectWriter capture enough state for the block uploader to resume mid-volume? Or does it start over from the beginning?

Author: Resume is not supported for both fs data mover and block data mover. That is, once break, the data mover needs to start from the beginning.
Checkpoint() on ObjectWriter could help to save time for fs data mover even though the backup still starts from beginning, but it is not enabled for Velero, because it increases the complexity.
Checkpoint() will not be supported for block data mover, the implementation is totally different, but since it is not enabled for Velero, we will not implement it for the first implementation.

In any case, the ring buffer implementation won't affect the resume feature, on resume, the data in the ring buffer is naturally lost, but the resume point is at the point before the data in the ring buffer.

To improve performance, block device is opened with direct IO, so that no data is going through the system cache unnecessarily.  

> During restore, to optimize the write throughput and storage usage, zero blocks should be either skipped (for restoring to a new volume) or unmapped (for restoring to an existing volume). To cover the both cases in a unified way, the SCSI command `WRITE_SAME` is used. Logics are as below:  

Reviewer 3: On the restore path, the design says the uploader reads blocks from Kopia, detects if they're all zeros, then issues BLKZEROOUT instead of writing. But for a 1TB volume with 1MB of real data, the uploader is still reading ~1TB of zero data from Kopia just to detect it's zero and skip the write. Can the uploader know from the BAT/snapshot metadata which chunks are zero and skip reading them entirely? That would make restore I/O proportional to actual data size, same as backup.

Author: > But for a 1TB volume with 1MB of real data, the uploader is still reading ~1TB of zero data from Kopia just to detect it's zero and skip the write.

From the repository perspective, reading zero data is equivalent to reading metadata only, so it is very economic, no data is actually processed.

Author: > Can the uploader know from the BAT/snapshot metadata which chunks are zero and skip reading them entirely? 

Uploader doesn't know it, for many reasons BAT is kept by the repository internally. If the uploader really needs to know it, the repository should expose interfaces for it. But as mentioned above, it is not actually required.

Uploader implementation is OS dependent, but since Windows container doesn't support block volumes, the current implementation is for linux only.  

#### ChangeId
ChangeId identifies the base that CBT is generated from, it must strictly map to the parent snapshot in the repository. Otherwise, there will be data corruption in the incremental backup.  
Therefore, ChangeId is saved together with the repository snapshot.  
The data mover always queries parent snapshot from Unified Repo together with the ChangeId. In this way, no mismatch would happen.  
Inside the uploader, the upper layer (DataUpload controller) could also provide the ChangeId as a mechanism of double confirmation. The received ChangeId would be re-evaluated against the one in the provided snapshot.    

For Kubernetes API, changeId is represented by `BaseSnapshotId`.  
changeId retrieval is storage specific, generally, it is retrieved from the `SnapshotHandle` of the VolumeSnapshotContent object; however, storages may also refer to other places to retrieve the changeId.  
That is, `SnapshotHandle` and changeId may be two different values, in this case, the both values need to be preserved.  

> #### Volume Snapshot retention

Reviewer 3: Two things about Volume Snapshot Retention:                                                                                                                              
                                                                                                                                                                         
**RetainSnapshot only works with Volume Policy**: if a user just does `velero backup create --data-mover velero-block` without any Volume Policy, there's no way to set RetainSnapshot. They're stuck with Way 1 (delete after backup). If their storage needs the base snapshot to exist for GetMetadataDelta, every incremental silently falls back to full. Should we expose this at the Backup spec or BSL level too?                                                                                                
                                                                                                                                                                         
**Silent fallback**: when incremental falls back to full (CBT error, missing parent, etc.), how does the user know? They could run scheduled "incremental" backups for weeks without realizing every one is actually full because their retention config is wrong. Can we add a Warning event on the Backup CR and surface the requested vs actual type in `velero backup describe`?

Author: > RetainSnapshot only works with Volume Policy: if a user just does velero backup create --data-mover velero-block without any Volume Policy, there's no way to set RetainSnapshot. They're stuck with Way 1 (delete after backup). If their storage needs the base snapshot to exist for GetMetadataDelta, every incremental silently falls back to full. Should we expose this at the Backup spec or BSL level too?

I think we will stick to the Volume Policy only, reasons:
1. Retain snapshot is a non-foundamental feature
2. As our evolvement plan, we want to add new configurations to configMaps instead of the CRs or CLI
3. From the nature of this feature, it requires a fine grained configuration, e.g., configure it by storageclass/CSI driver, simply parameters in CR/CLI cannot fulfill it

Author: > Silent fallback: when incremental falls back to full (CBT error, missing parent, etc.), how does the user know? They could run scheduled "incremental" backups for weeks without realizing every one is actually full because their retention config is wrong. Can we add a Warning event on the Backup CR and surface the requested vs actual type in velero backup describe?

Yes, users will be able to see it through backup describe, see this statement in the design:
```
Backup type should be added to backup description, there are two appearances:
- The `backupType` in the Backup CR. This is the selected backup type by users
- The backup type recorded in `volumeInfo.json`, which is the actual type taken by the backup
With these two values, users are able to know the actual backup type and also whether a fallback happens. 
```

#### Incremental Size
By the end of the backup, incremental size is also returned by the uploader, as same as Velero file system uploader. The size indicates how much data are unique so processed by the uploader, based on the provided CBT.  

### Fallback to Full Backup
There are some occasions that the incremental backup won't continue, so the data mover fallbacks to full backup:
- `GetMetadataAllocated` or `GetMetadataDelta` returns error
- ChangeId is missing
- Parent snapshot is missing

### Irregular Volume Size
As mentioned above, during incremental backup, block uploader IO should be restricted to be aligned to the deduplication chunk size (1MB); on the other hand, there is no hard limit for users' volume size to be aligned.  
To support volumes with irregular size, below measures are taken:
- Volume objects in the repository is always aligned to 1MB
- If the volume size is irregular, zero bytes will be padded to the tail of the volume object
- A real size is recorded in the repository snapshot
- During restore, the real size of data is restored

The padding must be always with zero bytes.  

> ### Volume Size Change

Reviewer 3: Does the CBT API guarantee correct deltas when a volume was resized between two snapshots? If snapshot A was taken at 1.5GB and snapshot B at 2.5GB, does GetMetadataDelta(A, B) return valid results across all storage backends? If a storage doesn't handle this correctly in its delta computation, the uploader would write wrong data, that's not a fallback-to-full situation, that's silent data corruption.

Author: `GetMetadataDelta` supports the volume resize case, so if data is really changed in the resized area, it will report the changes.
However, if the resize area is not changed, e.g., volume is extended but no data is actually written; or volume is shrunk, it doesn't report it.

This is why the uploader needs to have its own logics to handle the resize case, instead of relying on CBT.

- Otherwise, if volume is expanded, `WriteAt` guarantees to clone appropriate objects entries from the parent object and append zero data for the expanded areas. Particularly, if the parent volume is not in regular size, the zero padding bytes is also reused. Therefore, the parent object's padding bytes must be zero
- In the case the volume is shrunk, writing the tail data makes sure zero bytes are padding to the new volume object instead of inheriting non-zero data from the parent object

### Cancellation
The existing Cancellation mechanism is reused, so there is no change outside of the block uploader.  
Inside the uploader, cancellation checkpoints are embedded to the uploader reader and writer, so that the execution could quit in a reasonable time once cancellation happens.  

### Parallelism
Parallelism among data movers will reuse the existing mechanism --- load concurrency.  
Inside the data mover, uploader reader and writer are always running in parallel. The number of reader and writer is always 1.  
Sequential read/write of the volume is always optimized, there is no prove that multiple readers/writers are beneficial.  

> ### Progress Report

Reviewer 3: For incremental backups, is TotalBytes in the progress report the full volume size or just the changed bytes? If it's the full volume and only 10MB changed on a 1TB volume, progress jumps from 0 to ~99% immediately which isn't useful. If it's the changed bytes, progress is meaningful. Worth specifying.

Author: `TotalBytes` reports the full volume size, so the progress only reflects the logical progerss, this is aligned with fs data mover. E.g., for a 1TB file system, incremental backup eliminates 99% of unchanged files, but the progress still reflects the 1TB progress, even though it is very fast.

And the same as fs data mover, by then end, the uploader to call the progress's `GetIncrementalSize` to reflect how much data has been eliminated because of incremental backup. I have updated the same to the design.

#### File System Data Mover
At present, Velero file system data mover doesn't support selectable backup type, instead, incremental backups are always conducted once possible.  
From user experience this is not reasonable.  

Therefore, to solve this problem and to make it align with Velero block data mover, Velero file system data mover will support backup type as well.  

At present, the data path for Velero file system data mover has already supported it, we only need to expose this functionality to users.  

### Backup Describe
Backup type should be added to backup description, there are two appearances:
- The `backupType` in the Backup CR. This is the selected backup type by users
- The backup type recorded in `volumeInfo.json`, which is the actual type taken by the backup
With these two values, users are able to know the actual backup type and also whether a fallback happens.  

The `DataMover` item in the existing backup description should be updated to reflect the actual data mover completing the backup, this information could be retrieved from `volumeInfo.json`.  

### Backup Sync
No more data is required for sync, so Backup Sync is kept as is.  

### Backup Deletion

> As mentioned above, no data is moved when deleting a repo snapshot for Velero block data mover, so Backup Deletion is kept as is.  

Reviewer 3: Backup Deletion section says "no data is moved when deleting a repo snapshot, so Backup Deletion is kept as is." But the Volume Snapshot Retention section describes new deletion logic for the RetainSnapshot case, rebuilding a VS/VSC pair with deletionPolicy=Delete and deleting it to clean up the storage snapshot. These two sections seem to contradict each other. Can the Backup Deletion section acknowledge the RetainSnapshot cleanup path?

Author: Yes, for Volume Snapshot retention case, backup deletion needs to be modified.
I have updated the design accordingly.

### Restarts
Restarts mechanism is reused without any change.  

### Logging
Logging mechanism is not changed.  

> ### Backup CRD

Reviewer 3: Default backupType is "incremental" but the first backup for a volume will always have no parent and fall back to full. Should the controller detect "no previous backup for this volume" and set it to full automatically, rather than having it show as incremental that fell back to full? Small thing but cleaner from a user perspective.

Author: It looks more comfortable to have the first backup as full, because the backup describe will report a backup method switch (it can gives a better report than the fallback warning, if we process the case specially and smarter). However, in another word, if we process the case smarter in backup describe, it is not necessary to set the backup as full.
Actually, many solutions don't allow users to set the backup method, full/incremental, because those solutions believes in the incremental-forever manner very much.

And here we also leave it to retention policy, which has more configurations for specifying which one is full and which one is incremental.
This is why Velero supports to configure a full backup from UX level.

Finally, from the implementation level:
1. It is not good enough for controller to detect and switch the method, because any internal processing switches the method, there should be a report. So in terms of this, there is no difference for controller to do the switch or uploader does this
2. Controller cannot do it, because for accurately detecting it, repo must be connected, while we are avoiding to connect to the repo outside of the data mover, because repo connection may be a resource consuming operation

Author: Therefore, we can give a better and special treatment for this case in backup describe, e.g., in backup describe we can say "automatically switch to full for the first backup" instead of gives a warning like fallback happens.
